tests/integration/test_engine_integration.py

```diff
@@ -126,7 +126,7 @@ def test_return_samples(self, setup_eng):
         assert [r.val for r in prog.register] == res.samples[0].tolist()
         # first mode was measured
         if eng.backend_name == 'tf':
-            assert isinstance(res.samples[0], tf.Tensor)
+            assert isinstance(res.samples[0][0], tf.Tensor)
```

Comment:
`res.samples[0]` is returning the following in this specific test:
```python
>>> res.samples[0]
array([[<tf.Tensor: id=520, shape=(), dtype=float64, numpy=0.02130021300212981>
  None]], dtype=np.object)
```

Is this unexpected?

Reply:
I think all the backend methods should return consistent results, i.e. the dimensionality should not depend on the backend, but this is clearly a question for another PR.

Reply:
@thisac, perhaps this is a bug in the new results shape processing when receiving TensorFlow tensors(?)

Reply:
That could definitely be the case. I'll look into it tomorrow!

Reply:
Everyone always seems to forget, but the TF backend supports a special case that the other backends do not support, namely batching. The tests were originally written so that they pass with or without batching, which sometimes results in unusual-looking code

Reply:
I looked into this and it seems like, in `engine.py`, the samples are transformed into an array and sent to the `Result` class where the samples are reshaped into `(shots, modes)` by adding `np.array([self.samples])`.

Each value in `self.samples` is either a float or a `tf.Tensor` or `None`. So, what the test above is comparing is:

1. the first value of the first shot in `self.samples` if using the `tf` backend (which should be a `tf.Tensor` if using `tf` and a float otherwise) or

2. the first shot in `self.samples` if using other backends (which should _always_ be an `ndarray`, even for the `tf` backend).

So, I don't think there's any consistency faults here really, rather the `tf` test checks the type of _one_ value in `self.samples` while the non-`tf` test checks the type of one shot (i.e. an array several values).

Should the `tf` backend return samples with the type `ndarray[tf.Tensor]` and shape (shots, modes)? In that case I think that this is actually OK. I guess the alternative would be to return a `tf.Tensor[float]` with shape `(shots, modes)`, but I'm not sure how/where that should be changed or if that is actually wanted.

Reply:
Definitely the second option is preferred. The tensorflow backend should return a tensor tensor with the same size/shape as the comparable numpy array. In the case of batched mode, there should be an extra first axis (but batched mode is not co-compatible with many-shots measurement, so that shouldn't be a problem in this specific case)

Wrapping a tensorflow tensor in a numpy array is unnecessary, and may also break the computational graph unnecessarily.

Reply:
I agree that the second option sounds nicer. Although right now the samples are set to `None` whenever there are no measurements, meaning that the samples might be an `ndarray[object]` with some floats or `tf.Tensor`s (depending on backend) and some `Nones`. Looking at the comment above from @josh146, `res.samples` is just that (an `ndarray[object]` with one `tf.Tensor` and one `None`) and, since I don't think that `tf.Tensor` supports `None` objects further changes might need to be done in the process to make everything work.

Actually, I also noticed that right now, if the tensorflow backend is used with a single shot (where all modes are measured), `engine.py` will actually cast `self.samples` into a numpy array (i.e. removing the tensorflow tensor altogether). I'm guessing this is not desirable.

Reply:
The `None`s were just a placeholder and may not make sense anymore. Feel free to suggest a strategy to replace them

Reply:
> Actually, I also noticed that right now, if the tensorflow backend is used with a single shot (where all modes are measured), `engine.py` will actually cast `self.samples` into a numpy array (i.e. removing the tensorflow tensor altogether). I'm guessing this is not desirable.

Again, I think this is an artifact of how we used to do things with the TF1 days. The default was for tensorflow tensors to be "evaluated", which turned them into numpy arrays. Users could specify a kwarg to force the tensors to remain "unevaluated" (i.e., abstract tensors objects with no set data value). Now that tensorflow has thrown out sessions, tensors are somehow always evaluated. When we need them to be specifically numpy objects, we use `.numpy()`